In [10]:
import pandas as pd

df = pd.read_csv('financial.csv')
df.head()


Company  Fiscal Year  Total Revenue  Net Income  Total Assets  \
0      Apple         2023         383285       96995        352583   
1      Apple         2022         394328       99803        351002   
2      Apple         2021         365817       94680        351002   
3  Microsoft         2023         232315       72361        411987   
4  Microsoft         2022         198270       72738        364840   

   Total Liabilities  Cash Flow from Operating Activities  
0             267645                               110543  
1             278829                               122151  
2             287912                               104038  
3             222515                                95152  
4             198298                                89034

In [11]:
# Convert financial columns to numeric values (force errors to NaN if they exist)
cols_to_convert = [
    'Total Revenue', 'Net Income', 'Total Assets',
    'Total Liabilities', 'Cash Flow from Operating Activities'
]

for col in cols_to_convert:
    df[col] = pd.to_numeric(df[col], errors='coerce')


In [12]:
# Calculate ratios
df['Net Profit Margin'] = df['Net Income'] / df['Total Revenue']
df['ROA'] = df['Net Income'] / df['Total Assets']
df['Debt to Asset'] = df['Total Liabilities'] / df['Total Assets']
df['OCF Margin'] = df['Cash Flow from Operating Activities'] / df['Total Revenue']


In [13]:
df['Revenue Growth (%)'] = df.groupby('Company')['Total Revenue'].pct_change() * 100
df['Net Income Growth (%)'] = df.groupby('Company')['Net Income'].pct_change() * 100


In [14]:
df = df.round({'Revenue Growth (%)': 2, 'Net Income Growth (%)': 2})
print(df)


     Company  Fiscal Year  Total Revenue  Net Income  Total Assets  \
0      Apple         2023         383285       96995        352583   
1      Apple         2022         394328       99803        351002   
2      Apple         2021         365817       94680        351002   
3  Microsoft         2023         232315       72361        411987   
4  Microsoft         2022         198270       72738        364840   
5  Microsoft         2021         168088       61271        333779   
6      Tesla         2023          96773       15001        102664   
7      Tesla         2022          81462       12556         82338   
8      Tesla         2021          53823        5519         62131   

   Total Liabilities  Cash Flow from Operating Activities  Net Profit Margin  \
0             267645                               110543           0.253062   
1             278829                               122151           0.253096   
2             287912                               104038  

In [9]:
df['Company'] = df['Company'].ffill()


## 📊 Key Insights from Financial Analysis (2021–2023)

### 🧾 Companies Analyzed
- Apple
- Microsoft
- Tesla

---

### 📈 Revenue Growth (2021 → 2023)
- **Apple**: Peaked in 2022, but slight **decline in 2023** (−2.8%). Growth has flattened.
- **Microsoft**: Strong upward trend in 2023 (**+17.2%** from 2022), reversing previous declines.
- **Tesla**: Highest growth among the three (**+79.7%** from 2021 to 2023), but slowing down after 2022.

---

### 💰 Net Income Trends
- **Apple**: Fairly stable net income across years, but **slight dip** in 2023.
- **Microsoft**: Minor fluctuations; steady profits.
- **Tesla**: Massive improvement—from **$5.5B in 2021 to $15B in 2023** (over 170% growth).

---

### 📉 Net Profit Margin (2023)
- **Microsoft**: Highest at **31.1%**, very strong profitability.
- **Apple**: Solid at **25.3%**.
- **Tesla**: Lower margin at **15.5%**, but still improving from earlier years.

---

### 🏦 Return on Assets (ROA)
- **Apple**: Best ROA at **27.5%**—high asset efficiency.
- **Microsoft**: Decent at **17.5%**.
- **Tesla**: Much lower at **14.6%**, but improving.

---

### 🧮 Financial Health
**Debt to Asset Ratio (2023):**
- **Tesla**: Lowest at **0.37** – most conservative debt structure.
- **Microsoft**: Moderate at **0.54**.
- **Apple**: High at **0.76** – more leveraged, possibly to fund buybacks or R&D.

---

### 🔄 Cash Flow from Operating Activities (2023)
- **Apple**: $110.5B
- **Microsoft**: $95.1B
- **Tesla**: $13.9B  
Despite Tesla’s growth, it still trails in absolute cash flow due to scale.

---

### 🧠 Final Takeaways
- **Apple** remains a highly profitable and efficient company, but its growth has plateaued.
- **Microsoft** is in a strong position with consistent margins, solid growth, and moderate debt.
- **Tesla** is the fastest-growing, with improving fundamentals, but still behind in profitability and scale compared to the other two.

